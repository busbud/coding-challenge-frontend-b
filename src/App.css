body {
    font-family: 'Open Sans Condensed';
}

a {
    text-decoration: none;
    color: black;
}

* {
    box-sizing: border-box;
}