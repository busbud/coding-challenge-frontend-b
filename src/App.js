import React, { Fragment } from 'react';
import SearchForm from './components/searchForm/SearchForm';
import Departures from './components/departures/Departures';
import Header from './components/header/Header';

const App = () => {
    return (
        <Fragment>
            <Header />
            <SearchForm />
            <Departures />
        </Fragment >
    )
}
export default App;