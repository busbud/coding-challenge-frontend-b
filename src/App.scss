// Style for main pages
main {
  padding-top: 10px;
  margin-left: auto;
  margin-right: auto;
  max-width: 900px;
}
