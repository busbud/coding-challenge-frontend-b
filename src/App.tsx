import { Suspense, useEffect, useRef, useState, useMemo } from "react";
import i18next from "i18next";
import { initReactI18next, useTranslation } from "react-i18next";
import { Router, Route, Switch } from "react-router-dom";
import { createBrowserHistory } from "history";
import routes from "./routes/routes";
import { languages } from "utils/languages";
import {
  BusbudContextValues,
  BusbudContext,
} from "./interfaces/context.interface";
import { handleNavClick } from "utils/utils";

// Initialize the History Browser History.
const history = createBrowserHistory();

// Check Local Store Language value.
const lang = () => {
  return localStorage.lang
    ? localStorage.lang
    : localStorage.setItem("lang", "en");
};

// Initialze i18next.
i18next.use(initReactI18next).init({
  resources: languages,
  lng: lang(),
  fallbackLng: "en",
  interpolation: {
    escapeValue: false,
  },
});

const App: React.FC = () => {
  const langRef = useRef<HTMLSelectElement>(null);
  const [langValue, setLangValue] = useState(localStorage.lang);
  const [currencyValue, setCurrencyValue] = useState("CAD");

  const { t } = useTranslation();

  useEffect(() => {
    lang();
  }, []);

  // Set Context values to be accessible throughout the entire application.
  const values: BusbudContextValues = useMemo(
    () => ({
      currencyValue,
    }),
    [currencyValue]
  );

  return (
    <BusbudContext.Provider value={values}>
      <div className="App">
        <nav className="nav">
          <div className="nav-container">
            <h1 onClick={handleNavClick("/", history)}>BusBud</h1>
            <div className="nav-container-selectors">
              <select
                ref={langRef}
                name="lang"
                id="lang"
                value={langValue}
                onChange={(e) => {
                  i18next.changeLanguage(e.target.value);
                  localStorage.setItem("lang", e.target.value);
                  setLangValue(e.target.value);
                }}
              >
                <option value="en">English</option>
                <option value="fr">French</option>
              </select>
              <select
                name="currency"
                id="currency"
                value={currencyValue}
                onChange={(e) => {
                  setCurrencyValue(e.target.value);
                }}
              >
                <option value="CAD">CAD</option>
                <option value="USD">USD</option>
                <option value="EUR">EUR</option>
              </select>
            </div>
          </div>
        </nav>
        <article className="body">
          <Suspense fallback={<span>Loading...</span>}>
            <Router history={history}>
              <Switch>
                {routes(t).map((route, key) => (
                  <Route
                    key={key}
                    path={route.path}
                    component={route.component}
                    exact={route.exact}
                  />
                ))}
              </Switch>
            </Router>
          </Suspense>
        </article>
        <footer className="footer">
          <div className="footer-container">
            &copy; BusBud {new Date().getFullYear()}
          </div>
        </footer>
      </div>
    </BusbudContext.Provider>
  );
};

export default App;
