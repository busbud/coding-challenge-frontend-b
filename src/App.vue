<template>
  <div id="app">
    <b-navbar toggleable="md" type="dark" id="navbar">
      <b-navbar-toggle target="nav_collapse"></b-navbar-toggle>
      <b-navbar-brand href="#">BusBud</b-navbar-brand>
      <b-collapse is-nav id="nav_collapse">
        <!-- Right aligned nav items -->
        <b-navbar-nav class="ml-auto">
          <b-nav-item-dropdown text="Lang" right>
            <b-dropdown-item href="#">EN</b-dropdown-item>
            <b-dropdown-item href="#">FR</b-dropdown-item>
          </b-nav-item-dropdown>
        </b-navbar-nav>
      </b-collapse>
    </b-navbar>
    <div class="container">
      <img src="./assets/osheaga.png" id="osheaga-logo"/>
      <router-view/>
    </div>
  </div>
</template>

<script>
export default {
  name: 'App'
}
</script>

<style lang="scss">

$navbar-color: #f19020;
$navbar-margin: 25px;

#app {
  font-family: 'Avenir', Helvetica, Arial, sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  text-align: center;
  color: #2c3e50;
}
#navbar {
  background-color: $navbar-color;
}
#osheaga-logo {
  margin-top:$navbar-margin;
  margin-bottom:$navbar-margin;
  max-width:100%;
}
</style>
