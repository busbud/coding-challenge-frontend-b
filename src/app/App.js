import React, { Component } from 'react';
import { translate } from 'react-i18next';

import Footer from '../components/Footer';
import Header from '../components/Header';
import MainSection from '../components/MainSection';
import parseDepartures from '../api/parser';
import { initialFetch, poll } from '../api/service';
import { delay } from '../utils/utils';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      searchParams: {
        origin: 'dr5reg',
        destination: 'f25dvk',
        date: new Date(Date.UTC(2018, 7, 2)),
      },
      language: props.i18n.language,
      isLoading: false,
      isMenuActive: false,
      departures: [],
      error: null,
    };
  }

  componentDidMount() {
    this.startDeparturesFetch();
  }

  handleApiError = (err) => {
    const { t } = this.props;
    const msg = t('msg.error_api');
    console.error(msg, err);
    this.setState({
      error: msg,
      isLoading: false,
    });
  };

  handleLanguageChange = (langId) => {
    this.props.i18n.changeLanguage(langId);
    this.setState({
      language: langId,
      isMenuActive: false,
    });
  };

  handleMenuClick = () => (
    this.setState(prevState => ({ isMenuActive: !prevState.isMenuActive }))
  );

  handleSearchClick = (origin, destination, date) => {
    this.setState({
      searchParams: { origin, destination, date },
      isMenuActive: false,
      departures: [],
      error: null,
    }, () => this.startDeparturesFetch());
  };

  startDeparturesFetch() {
    this.setState({ isLoading: true });

    initialFetch(this.state.searchParams).then((initialData) => {
      console.info('initial fetch has completed.', initialData);

      this.setState({
        departures: parseDepartures(initialData),
      });

      if (!initialData.complete) {
        return this.pollDepartures(5);
      }

      return this.setState({ isLoading: false });
    }).catch(this.handleApiError);
  }

  pollDepartures(iterations) {
    let countdown = iterations;

    return poll({
      ...this.state.searchParams,
      index: this.state.departures.length,
    }).then((newData) => {
      console.info(`poll #${countdown} has completed.`, newData);
      this.setState({
        departures: this.state.departures.concat(parseDepartures(newData)),
      });

      countdown -= 1;
      if (countdown > 0 && !newData.complete) {
        return delay(2000).then(() => this.pollDepartures(countdown));
      }
      return this.setState({ isLoading: false });
    });
  }

  render() {
    return (
      <div className="App">
        <Header
          currentLang={this.state.language}
          isMenuActive={this.state.isMenuActive}
          searchParams={this.state.searchParams}
          onLanguageClick={this.handleLanguageChange}
          onMenuClick={this.handleMenuClick}
          onSearchClick={this.handleSearchClick}
        />
        <MainSection
          currentLang={this.state.language}
          currentSearch={this.state.searchParams}
          departures={this.state.departures}
          error={this.state.error}
          isLoading={this.state.isLoading}
        />
        <Footer />
      </div>
    );
  }
}

export default translate()(App);
