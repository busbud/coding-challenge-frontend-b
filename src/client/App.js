import React, { Component } from 'react';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    return <div>Hello World</div>;
  }
}

export default App;
