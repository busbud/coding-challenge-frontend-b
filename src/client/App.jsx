import React, { Component } from 'react';
import MainContainer from './containers/MainContainer';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    return <MainContainer />;
  }
}

export default App;
