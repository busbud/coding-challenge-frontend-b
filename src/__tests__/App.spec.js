import React from 'react';
import App from '../App';
import renderer from 'react-test-renderer';
jest.mock('../components/searchForm/SearchForm', () => 'SearchForm');
jest.mock('../components/departures/Departures', () => 'Departures');

it('renders correctly', () => {
  const tree = renderer
    .create(<App />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});