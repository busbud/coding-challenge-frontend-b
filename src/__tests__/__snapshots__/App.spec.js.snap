// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <div
    className="lang"
  >
    <div
      className="lang"
    >
      <div
        onClick={[Function]}
      >
        ğŸ‡¬ğŸ‡§
      </div>
      <div
        onClick={[Function]}
      >
        ğŸ‡«ğŸ‡·
      </div>
    </div>
  </div>,
  <SearchForm />,
  <Departures />,
]
`;
